### PR TITLE
Fetch tools fixes

### PR DESCRIPTION
**Pull request by Orzelius**
please help yourself to the commits, they're rather chonky, sorry

### CODE REVIEW

**Comment by Orzelius** [comment]
some tests might fail, I'll fix in a separate pr

**Comment by Orzelius** [comment]
Fixed the tests. Ready for review. @Walther Would this collide with #4381?

File: core/src/commands/util/fetch-tools.ts [hunk 1]
```diff
@@ -90,10 +94,24 @@ export class FetchToolsCommand extends Command<{}, FetchToolsOpts> {
...
     const deduplicated = uniqBy(tools, ({ tool }) => tool["versionPath"])
 
+    const versionedConfigs = garden.getRawProviderConfigs({ names: ["pulumi", "terraform"], allowMissing: true })
+
+    // If the version of the tool is configured on the provider,
+    // download only that version of the tool.
+    const toolsNeeded = deduplicated.filter(tool => {
+      const pluginToolVersion = versionedConfigs.find(p => p.name === tool.plugin.name)?.version
+      const pluginHasVersionConfigured = !!pluginToolVersion
+      if (!pluginHasVersionConfigured) {
```
Comment (vvagaytsev) [inline core/src/commands/util/fetch-tools.ts:104]: This looks a bit awkward, how about one line like `if (!pluginToolVersion)`?

File: core/src/util/util.ts [hunk 2]
```diff
@@ -521,11 +521,21 @@ export function pickKeys<T extends object, U extends keyof T>(obj: T, keys: U[],
...
+export function findByNames<T extends ObjectWithName>({
+  names,
+  entries,
+  description,
+  allowMissing = false,
+}: {
+  names: string[]
+  entries: T[]
+  description: string
+  allowMissing?: boolean
```
Comment (vvagaytsev) [inline core/src/util/util.ts:533]: It looks like `allowMissing` can be non-optional without the default `false` value. The value is always ensured by the `getRawProviderConfigs`.

File: core/src/commands/util/fetch-tools.ts [hunk 3]
```diff
@@ -90,10 +94,24 @@ export class FetchToolsCommand extends Command<{}, FetchToolsOpts> {
     // No need to fetch the same tools multiple times, if they're used in multiple providers
     const deduplicated = uniqBy(tools, ({ tool }) => tool["versionPath"])
 
+    const versionedConfigs = garden.getRawProviderConfigs({ names: ["pulumi", "terraform"], allowMissing: true })
```
Comment (vvagaytsev) [inline core/src/commands/util/fetch-tools.ts:97]: Do we need to check only `terraform` and `pulumi` providers here?

**Review by vvagaytsev — APPROVED** [review]
Looks good, thank you! 👍
Left a few non-blocking comments, please check.

File: core/src/commands/util/fetch-tools.ts (hunk 1 above)
Comment (Orzelius) [reply core/src/commands/util/fetch-tools.ts:104]: You mean the variable name? I chose it especially so it's not confused with the version of the tools available and the version that's configured on the plugin

Comment (vvagaytsev) [reply core/src/commands/util/fetch-tools.ts:104]: The name is good, I meant the variable itself. But, tbh, the extra variable here brings more clarity 👍

**Review by vvagaytsev — APPROVED** [review]
🚢

**Review by Walther — APPROVED** [review]
nice work, thank you!

this shouldn't conflict with the other pr (which has been merged now)

**Pull request merged**